pkg_auto/impl/md5_cache_lib.sh

```diff
@@ -0,0 +1,1177 @@
+#!/bin/bash
+
+# This file implements parsing the md5-metadata cache files and
+# accessing the parsed results. Not the entirety of the cache file is
+# parsed, only the parts that were needed at the time of writing
+# it. So currently the exposed parts of parsed cache files are EAPI,
+# IUSE, KEYWORDS, LICENSE, {B,R,P,I,}DEPEND and _eclasses_. The
+# _eclasses_ part discards the checksums, though, so only names are
+# available.
+
+if [[ -z ${__MD5_CACHE_LIB_SH_INCLUDED__:-} ]]; then
+__MD5_CACHE_LIB_SH_INCLUDED__=x
+
+source "$(dirname "${BASH_SOURCE[0]}")/util.sh"
+source "${PKG_AUTO_IMPL_DIR}/debug.sh"
+source "${PKG_AUTO_IMPL_DIR}/gentoo_ver.sh"
+
+#
+# Cache file
+#
+
+# Indices to access fields of the cache file:
+# PCF_EAPI_IDX     - a string/number describing EAPI
+# PCF_KEYWORDS_IDX - a name of an array containing keyword objects
+# PCF_IUSE_IDX     - a name of an array containing iuse objects
+# PCF_BDEPEND_IDX  - a name of a group object with build dependencies
+# PCF_DEPEND_IDX   - a name of a group object with dependencies
+# PCF_IDEPEND_IDX  - a name of a group object with install dependencies
+# PCF_PDEPEND_IDX  - a name of a group object with post dependencies
+# PCF_RDEPEND_IDX  - a name of a group object with runtime dependencies
+# PCF_LICENSE_IDX  - a name of a group object with licenses
+# PCF_ECLASSES_IDX - a name of an array with used eclasses
+declare -gri PCF_EAPI_IDX=0 PCF_KEYWORDS_IDX=1 PCF_IUSE_IDX=2 PCF_BDEPEND_IDX=3 PCF_DEPEND_IDX=4 PCF_IDEPEND_IDX=5 PCF_PDEPEND_IDX=6 PCF_RDEPEND_IDX=7 PCF_LICENSE_IDX=8 PCF_ECLASSES_IDX=9
```

Comment:
How sure are you that this order is fixed?

Reply:
I'm not sure if this is what you are asking about, but this order is not an order of entries showing up in the cache file on disk. These indices are for accessing `cache_file` members. If I wrote this in C, then the `cache_file` could be:

```C
struct cache_file {
    int EAPI;
    struct keyword* keywords;
    int keywords_len;
    struct iuse* iuses;
    int iuses_len;
    …
};
struct keywords kw[2] = {…};
struct iuse iu[5] = { … }
struct cache_file cf = { 8, kw, 2, iu, 5, … }:
…
printf("EAPI: %d\n", cf.EAPI);
```

Since it's bash, I implemented the struct as an array:

```bash
declare -a kw=( … )
declare -a iu=( … )
declare -a cf=( 8 kw ia … )
…
printf 'EAPI: %d\n' "${cf[PCF_EAPI_IDX]}"
```